diff --git a/deps/quickjs/quickjs.c b/deps/quickjs/quickjs.c
index d939308..9bd8f00 100644
--- a/deps/quickjs/quickjs.c
+++ b/deps/quickjs/quickjs.c
@@ -57170,6 +57170,11 @@ bool JS_IsDate(JSValueConst v)
     return JS_VALUE_GET_OBJ(v)->class_id == JS_CLASS_DATE;
 }
 
+int JS_GetDateValue(JSContext *ctx, JSValueConst v, double *epoch_ms)
+{
+    return JS_ThisTimeValue(ctx, epoch_ms, v);
+}
+
 int JS_AddIntrinsicDate(JSContext *ctx)
 {
     JSValue obj;
diff --git a/deps/quickjs/quickjs.h b/deps/quickjs/quickjs.h
index 552ebfb..e3ecf88 100644
--- a/deps/quickjs/quickjs.h
+++ b/deps/quickjs/quickjs.h
@@ -949,6 +949,7 @@ JS_EXTERN JSValue JS_NewProxy(JSContext *ctx, JSValueConst target,
 
 JS_EXTERN JSValue JS_NewDate(JSContext *ctx, double epoch_ms);
 JS_EXTERN bool JS_IsDate(JSValueConst v);
+JS_EXTERN int JS_GetDateValue(JSContext *ctx, JSValueConst v, double *epoch_ms);
 
 JS_EXTERN JSValue JS_GetProperty(JSContext *ctx, JSValueConst this_obj, JSAtom prop);
 JS_EXTERN JSValue JS_GetPropertyUint32(JSContext *ctx, JSValueConst this_obj,
